test/unit/math/prim/mat/fun/append_array_new_test.cpp

```diff
@@ -0,0 +1,213 @@
+#include <stan/math/prim/mat.hpp>
+#include <gtest/gtest.h>
+
+using namespace Eigen;
+
+/**
+ * Generate a new random variable, cast it to type T1, and store it in z. n0
+ * is ignored in this specialization
+ *
+ * @tparam T1 Scalar type
+ * @param n0 Ignored
+ * @param z Output variable
+ */
+template<typename T1>
+void build(int n0, T1 &z) {
+  z = T1(rand());
+}
+
+/**
+ * Generate a new random numbers to fill z (casting them to type T1).
+ *
+ * If z is a column vector, first resize to have n0 rows
+ *
+ * If z is a row vector, first resize to have n0 columns
+ *
+ * If z is a matrix, first resize to have n0 rows and n0 columns
+ *
+ * @tparam T1 Scalar type
+ * @param n0 Dimension
+ * @param z Output variable
+ */
+template<typename T1, int R, int C>
+void build(int n0, Eigen::Matrix<T1, R, C>& z) {
+  int nrow = n0, ncol = n0;
+
+  if(R == 1)
+    nrow = 1;
+
+  if(C == 1)
+    ncol = 1;
+
+  z = Eigen::Matrix<T1, R, C>::Random(nrow, ncol);
```

Comment:
You can cut ou tall the temps and just use
```
Random(R == 1 ? 1 : n0, C == 1 ? 1 : n0);
```
But I'd forego all these builder things that pass in mutable references and just return a matrix from this function.

And I'd just have this delegate to `build(n0, n0)` anyway (and return a matrix).